bsdlib/README.rst

```diff
@@ -3,9 +3,131 @@
+
+The library implements the following standard BSD functions:
+   
+* socket
+* close
+* bind
+* listen
+* accept
+* connect
+* recvfrom
+* recv
+* read
+* sendto
+* send
+* write
+* getsockopt
+* setsockopt
+* fcntl
+* poll
+* inet_pton
+* inet_ntop
+* getaddrinfo
+* freeaddrinfo
+
+.. note::
+   The ``select`` function is not yet implemented.
+
+
+Library internals
+*****************
+
+The purpose of the BSD library is to communicate with the nRF9160 modem firmware.
+This is done by sending events and getting responses over the Inter Processor Communication peripheral (IPC).
+To facilitate data transfer between the application and the modem, payload data is set in the shared RAM.
+
+The modem contains a full IP and DTLS/TLS stack as well as GNSS.
+The library internally communicates over a Remote Procedure Call (RPC) protocol using the IPC and the shared RAM to utilize the modules inside the modem.
+The library provides address families, types, and protocols that dispatch the protocols to the right backend in the modem and thus communicates with this backend.
+
+The BSD library aims to provide a standard programming interface on top of all of these backends so that the application developer does not have to manage all intricate details of the low-level parameters passed back and forth between the application and the modem.
+
+The figure below shows a simplified BSD library architecture.
+
+.. figure:: doc/images/bsd_lib_architecture.svg
+   :alt: BSD library architecture diagram
+
+   BSD library architecture diagram
+
+Porting flexibility
+*******************
+
+The BSD library has been designed in such way that you can port it to any OS.
+Therefore, procedures that would normally require some kind of OS interaction, like going into sleep and IRQ reprioritization, have been moved out of the library to an open :file:`c` file in which OS-specific integration  can be implemented.
+For more information, see :ref:`bsdlib_ug_porting`.
+
+Limitations
+***********
+
+GNSS, AGPS, and Modem DFU are currently not part of the generic sockets.
+Therefore, operating on them using ``poll`` is not possible yet.
+This might change in future releases of the BSD library.
+
+PDN sockets only support ``open``, ``close``, ``setsockopt``, and ``getsockopt`` in order to configure a PDN with a APN name.
+Therefore, there is no data communication on the socket, and it is not pollable.
+
+Limitations on the cryptographic sockets:
+
+* 1 TLS socket
+* 2 DTLS sockets
+* TCP max packet size of 4096
+* TLS internal buffer size of 2.5 KB
+* Only one DTLS/TLS handshake can be done at the time.
+  No parallel handshake is possible.
+  
+Modem limitations on the built-in key storage:
+
+* Max number of assets is 16 (slots).
+* 1 PSK takes 1 slot.
+* 1 Identity takes 1 slot.
+* 1 certificate might span over multiple slots depending on the size.
+  Testing is needed to verify that the certificate of choice gets into the modem and can be used by referencing the security tag.
```

Comment:
gets stored in the modem?

Reply:
changed to:
'is correctly stored in the modem'